From 1456a5841c63cc49b0165420eb79b190b385dfb4 Mon Sep 17 00:00:00 2001
From: Michael Walle <michael.walle@kontron.com>
Date: Wed, 28 Aug 2019 17:47:06 +0200
Subject: [PATCH] mtd-utils: add optional offset parameter to flash_otp_dump

There are flashes where there are gaps between OTP regions and flashes
where the regions don't start at 0. Pass an optional offset parameter.

Signed-off-by: Michael Walle <michael.walle@kontron.com>
---
 misc-utils/flash_otp_dump.c | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/misc-utils/flash_otp_dump.c b/misc-utils/flash_otp_dump.c
index f0c0fb9..f1e1782 100644
--- a/misc-utils/flash_otp_dump.c
+++ b/misc-utils/flash_otp_dump.c
@@ -10,16 +10,18 @@
 #include <string.h>
 #include <errno.h>
 #include <sys/ioctl.h>
+#include <stdlib.h>
 
 #include <mtd/mtd-user.h>
 
 int main(int argc,char *argv[])
 {
-	int fd, val, i, offset, ret;
+	int fd, val, i, ret;
+	int offset = 0;
 	unsigned char buf[16];
 
-	if (argc != 3 || (strcmp(argv[1], "-f") && strcmp(argv[1], "-u"))) {
-		fprintf(stderr,"Usage: %s [ -f | -u ] <device>\n", PROGRAM_NAME);
+	if (argc <= 3 || (strcmp(argv[1], "-f") && strcmp(argv[1], "-u"))) {
+		fprintf(stderr,"Usage: %s [ -f | -u ] <device> [<offset>]\n", PROGRAM_NAME);
 		return EINVAL;
 	}
 
@@ -36,9 +38,22 @@ int main(int argc,char *argv[])
 		return errno;
 	}
 
+	if (argc >= 4) {
+		char *p;
+		offset = (off_t)strtoull(argv[3], &p, 0);
+		if (argv[3][0] == 0 || *p != 0) {
+			fprintf(stderr, "%s: bad offset value\n", PROGRAM_NAME);
+			return ERANGE;
+		}
+	}
+
+	if (lseek(fd, offset, SEEK_SET) == (off_t)-1) {
+		perror("lseek()");
+		return errno;
+	}
+
 	printf("OTP %s data for %s\n",
 			argv[1][1] == 'f' ? "factory" : "user", argv[2]);
-	offset = 0;
 	while ((ret = read(fd, buf, sizeof(buf)))) {
 		if (ret < 0) {
 			perror("read()");
-- 
2.11.0

